### PR TITLE
Added functionality for post_local SGD

### LINKED ISSUE #74556: [Reliability Improvement] Record the step # in the optimizer state of PostLocalSGDOptimizer
### 🚀 The feature, motivation and pitch

**Context:**
[PostLocalSGDOptimizer](https://github.com/pytorch/pytorch/blob/8b0847811501cd452b131b6153f28c254a3ee44f/torch/distributed/optim/post_localSGD_optimizer.py#L5) can have a non-trivial warm-up stage. E.g., a training may take 30 epochs, and the warmup stage may take 3-5 epochs.

**Issue:**
If the training process is not very reliable and may fail a few times, then the training needs to resume from checkpoints. However, in the current setup the built-in [step counter](https://github.com/pytorch/pytorch/blob/8b8fac91bfca6cadf86d402622f31e8e2dcdc429/torch/distributed/algorithms/model_averaging/averagers.py#L22) in `ModelAverager` will be reset after each restart. As a result, the warm-up stage will run multiple times, which will lead to less speedup.

**Proposed Solution:**
To be able to skip the warm-up stage after resuming from a checkpiont, the step # across epochs must recorded and stored in the checkpoint as well. A good option is the optimizer state.

### Alternatives

If PyTorch Ignite or PyTorch Lightning is used, the global step # will be automatically stored in the checkpoint, and such feature can also be easily supported by modifying the application code.

However, it will be nice to support this feature in PyTorch directly, especially if an optimizer state is already available in this case.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by aovladi**
Fixes #74556

Added functionality to save and restore step counter for model averager.
Added a unittest.

### CODE REVIEW

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:76]: nit: There is no need for `.keys()` when checking key membership in a `dict`.
```suggestion
        if 'step' in state_dict:
```

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:70]: There is no need to check that `optim_state_dict is not None` since `torch.optim.Optimizer.state_dict()` never returns `None`.

Similarly, for `self.averager.step`, I do not believe it can be `None`.

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:81]: cc: @rohan-varma Is this the desired behavior? This seems reasonable to me.

If so, we should add a unit test that checks for a warning here (using `self.assertWarnsRegex(expected_warning=..., expected_regex=...)`. (The test loads a state dict without `"step"` as a top-level key and checks that the warning was issued.)

File: torch/testing/_internal/distributed/distributed_test.py
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:4940]: Is it possible to refactor the code common with `_test_post_localSGD_optimizer_parity()` into a helper method?

There is significant overlap.

File: torch/testing/_internal/distributed/distributed_test.py
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:4967]: See https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md for instructions on linting.

**Review by awgu — COMMENTED** [review]
Thanks for the quick PR! I am leaving some initial comments.

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (rohan-varma) [reply torch/distributed/optim/post_localSGD_optimizer.py:81]: yeah, we discussed this offline. 

might also be good to get @wayi1 who requested this feature feedback: If the saved state_dict did not contain a step, we're defaulting it to 0

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:4941]: is the file cleaned up at end of unittest? If not, consider using python tempfiles: https://docs.python.org/3/library/tempfile.html

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:4950]: I might miss something, but I can't quite tell how this comment relates to the below code?

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:4983]: possibly assert that it's nonzero to ensure we didn't hit the trivial case

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py:4940]: +1, or we can file a follow up issue

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:5046]: lets also have a unittest for the case we discussed offline - loading a checkpoint that does not have `step`.

For this you can probably take a checkpoint, remove `step` from the state dict, save that, reload, validate etc

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (rohan-varma) [reply torch/distributed/optim/post_localSGD_optimizer.py:81]: let's also add a unittest as @awgu suggested above (can be combined with my below unittest suggestion too). 

You can find documentation for assertWarnsRegex here: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarnsRegex

**Review by rohan-varma — COMMENTED** [review]
This is a great start and thanks for getting out the PR so quickly! Left a few comments

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (rohan-varma) [inline torch/distributed/optim/post_localSGD_optimizer.py:69]: Add a docstring mentioning it is the same as optim.Optimizer state_dict, but adds model averager's `step` to the checkpoint to ensure reload does not cause unnecessary warm up again?

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (rohan-varma) [inline torch/distributed/optim/post_localSGD_optimizer.py:76]: similar to above `state_dict` suggestion, can add a small docstring to `load_state_dict`

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (aovladi) [reply torch/distributed/optim/post_localSGD_optimizer.py:69]: good point, added this to an updated version

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (aovladi) [reply torch/distributed/optim/post_localSGD_optimizer.py:70]: got it, removed

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (aovladi) [reply torch/distributed/optim/post_localSGD_optimizer.py:76]: good point, removed from current version

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:3]: To avoid the CI failure, you can change this to `import warnings` and then call `warnings.warn()` in your code. I am not sure if there is a better way. cc: @rohan-varma

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (aovladi) [reply torch/distributed/optim/post_localSGD_optimizer.py:93]: Added unittest for warning check

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:70]: Mainly as an FYI: PyTorch follows [Google-style docstrings](https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html) and uses Sphinx to build the docs. This means that we can [cross-reference objects](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects) so that a reference here will link to that object's docs. (For example, if you go [here](https://pytorch.org/docs/stable/distributed.optim.html?highlight=zeroredundancyoptimizer#torch.distributed.optim.ZeroRedundancyOptimizer), you can click on `optim.Optimizer` to go to the `torch.optim.Optimizer` docs or click on `torch.nn.parallel.DistributedDataParallel` to go to the DDP docs.)

It can be a bit gimmicky, so I try to just check the docs created with the PR.

Some general notes:
- Double backticks (e.g. \`\`text\`\`) will render as code type (like `text`) but single backticks will not when you are not doing cross-references
- Cross-references only use single backticks with syntax like :meth:\`method_name\` or :class:\`class_name\`
- Docs will only render for public APIs and if there is a corresponding `.rst` file
- Try to use the cross-references like \`:class:torch.optim.Optimizer\` when appropriate since even if Sphinx cannot link the reference, it will render as code type

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:80]: Going off of the previous comment on the docstrings, I personally would write:
```suggestion
        This is same as :class:`torch.optim.Optimizer` `:meth:load_state_dict`,
```

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:82]: (continued)

Here, I like to double backtick the `state_dict` to make it clear it is referring to the argument. This can be useful sometimes if, for example, the argument is `module`, and it is not necessarily clear from the context if you mean module conceptually or the argument `module`.
```suggestion
        saved in provided ``state_dict``.
```

Disclaimer: I think I am the only one who is picky about these docstring things. Nonetheless, it is still good to be aware of it.

File: torch/testing/_internal/distributed/distributed_test.py
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:4988]: We would want to check `averager2.step != 0` instead of `not averager2.step is 0` since `is` checks for `int`s in Python may not be reliable: https://stackoverflow.com/questions/2239737/is-it-better-to-use-is-or-for-number-comparison-in-python

I think the change would be:

```suggestion
            self.assertNotEqual(averager2.step, 0)
```

However, should we just check the `.step` value to the desired positive integer (i.e. 20)? Otherwise, a passing implementation could simply set `step` to `1` if there is an entry in the state dict and `0` otherwise.

**Review by awgu — COMMENTED** [review]
I think the optimizer changes and unit test look good, but I am leaving some more comments, mostly style related. Feel free to re-request review after you have taken a look at the comments. We should be able to approve and land after that.

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (aovladi) [reply torch/distributed/optim/post_localSGD_optimizer.py:70]: Thanks for reference! Updated docstrings

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (aovladi) [reply torch/distributed/optim/post_localSGD_optimizer.py:80]: Thanks for the example!

File: torch/testing/_internal/distributed/distributed_test.py
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:4988]: Valid point. Addressed it.

File: torch/testing/_internal/distributed/distributed_test.py
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:5067]: Thanks for suggestion! Updated unittest accordingly

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (aovladi) [reply torch/distributed/optim/post_localSGD_optimizer.py:3]: Updated, will also take a note for this for future

File: torch/testing/_internal/distributed/distributed_test.py
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:4948]: Refactored common code

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (rohan-varma) [inline torch/distributed/optim/post_localSGD_optimizer.py:85]: nit: rise -> raise

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:4921]: spacing looks to be a bit off?

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:4959]: format looks to be off?

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:4964]: formatting again seems to be a bit strange here

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:5003]: nit: also assert that the loaded step is 0. Otherwise, a passing implementation could log that it is 0 and set it to some random value.

**Review by rohan-varma — APPROVED** [review]
LGTM overall, stamping to unblock. Please address the following before land:

- Lint / make sure CI lint jobs all pass
- Ensure all CI (other than unrelated flakiness) passes
- Misc comments / testing additions I mentioned
- Pls also get a stamp from @awgu

File: torch/testing/_internal/distributed/distributed_test.py
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:5011]: added on line 5013

**Review by rohan-varma — CHANGES_REQUESTED** [review]
Looks like the merge got botched.

You can probably discard unrelated changes in your working directory, then fetch latest origin and run `git rebase origin/master`.

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:81]: nit:
```suggestion
        but also restores model averager's step value to the one
```

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:82]: nit:
```suggestion
        saved in the provided ``state_dict``.
```

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:80]: nit:
```suggestion
        This is the same as :class:`torch.optim.Optimizer` :meth:`load_state_dict`,
```

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:84]: nit:
```suggestion
        If there is no ``"step"`` entry in ``state_dict``,
```

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:85]: nit:
```suggestion
        it will raise a warning and initialize the model averager's step to 0.
```

File: torch/distributed/optim/post_localSGD_optimizer.py
Comment (awgu) [inline torch/distributed/optim/post_localSGD_optimizer.py:91]: nit: I do not think you need a `+` for string concatenation between parentheses.

File: torch/testing/_internal/distributed/distributed_test.py
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:5008]: Nit: Same as above; no need for `+`.

**Review by awgu — APPROVED** [review]
I left some very minor nits. Looks good to me!

Let @rohan-varma  take a look as well.

**Review by rohan-varma — APPROVED** [review]
Looks good! Please address Andrew's comments and land away once CI is clear :D

**Comment by wayi1** [comment]
Thanks for the fix!

I believe this can lead to a much smoother adoption of industry users like Cruise, because it is very likely that industry users may have more training reliability issues than Meta, and hence it is important to allow post-localSGD to still work after retry.

**Pull request closed without merging**
